File: rdf-infographic-skill/scripts/rdf_parser.py
"""Parse RDF documents and extract KG Explorer data + narrative sections."""

from __future__ import annotations
import re
from pathlib import Path
from rdflib import Graph, URIRef, BNode, Literal
from rdflib.namespace import RDF, RDFS, OWL, XSD, Namespace


SCHEMA = Namespace("http://schema.org/")
PROV = Namespace("http://www.w3.org/ns/prov#")
KNOWN_CLASS_URIS = {
    RDF.Property, RDFS.Class, OWL.Class, OWL.NamedIndividual,
    SCHEMA.Person, SCHEMA.Organization, SCHEMA.Article,
    SCHEMA.FAQPage, SCHEMA.Question, SCHEMA.DefinedTermSet,
    SCHEMA.DefinedTerm, SCHEMA.HowTo, SCHEMA.HowToStep,
    SCHEMA.SoftwareApplication, SCHEMA.SoftwareSourceCode,
    SCHEMA.Thing, SCHEMA.CreativeWork,
}


def classify(node_uri: URIRef, g: Graph) -> str:
    """Classify a URIRef node as Class, Property, or Instance."""
    types = set(g.objects(node_uri, RDF.type))
    if not types:
        # Check if it's used as a predicate
        if (None, node_uri, None) in g or (node_uri, RDF.type, RDF.Property) in g:
            return "Property"
        # Check if it's used as a class
        for s, p, o in g:
            if p == RDF.type and o == node_uri:
                return "Class"
        return "Instance"

    for t in types:
        if t in (RDFS.Class, OWL.Class):
            return "Class"
        if t == RDF.Property:
            return "Property"

    for t in types:
        if t in KNOWN_CLASS_URIS:
            return "Instance"

    return "Instance"


def shorten(uri: URIRef, g: Graph) -> str:
    """Try to shorten a URI using namespace prefixes from the graph."""
    for prefix, ns in g.namespaces():
        if str(uri).startswith(str(ns)):
            return f"{prefix}:{str(uri)[len(str(ns)):]}"
    # Last resort: extract local name
    uri_str = str(uri)
    if "#" in uri_str:
        return uri_str.split("#")[-1]
    return uri_str.split("/")[-1] if "/" in uri_str else uri_str


def extract_label(uri: URIRef, g: Graph) -> str:
    """Extract the best label for a URI."""
    for label in g.objects(uri, RDFS.label):
        return str(label)
    for label in g.objects(uri, SCHEMA.name):
        return str(label)
    return shorten(uri, g)


def extract_description(uri: URIRef, g: Graph) -> str:
    """Extract description/comment for a URI."""
    for desc in g.objects(uri, RDFS.comment):
        return str(desc)[:200]
    for desc in g.objects(uri, SCHEMA.description):
        return str(desc)[:200]
    return ""


def build_kgdata(rdf_path: str | Path) -> dict:
    """Build kgData payload from an RDF file.

    Returns: {'nodes': [...], 'links': [...]}
    """
    g = Graph()
    g.parse(str(rdf_path))

    nodes_map: dict[str, dict] = {}
    links: list[dict] = []
    seen_predicates: set[str] = set()

    for s, p, o in g:
        if isinstance(s, BNode) or isinstance(o, BNode) and isinstance(p, URIRef):
            continue

        pred_short = shorten(p, g) if isinstance(p, URIRef) else str(p)
        seen_predicates.add(pred_short)

        subj_id = str(s) if isinstance(s, URIRef) else f"_:{s}"
        obj_id = str(o) if isinstance(o, URIRef) else f"_:{o}"

        # Add subject node
        if subj_id not in nodes_map and isinstance(s, URIRef):
            nodes_map[subj_id] = {
                "id": subj_id,
                "group": classify(s, g),
                "label": extract_label(s, g),
                "desc": extract_description(s, g),
                "iri": str(s),
            }

        # Add object node
        if obj_id not in nodes_map and isinstance(o, URIRef):
            nodes_map[obj_id] = {
                "id": obj_id,
                "group": classify(o, g),
                "label": extract_label(o, g),
                "desc": extract_description(o, g),
                "iri": str(o),
            }

        # Add link
        if isinstance(p, URIRef) and isinstance(s, (URIRef, BNode)) and isinstance(o, (URIRef, BNode)):
            link = {
                "source": subj_id,
                "target": obj_id,
                "predicate": pred_short,
                "label": pred_short,
            }
            links.append(link)

    nodes = list(nodes_map.values())

    # Orphan check
    incident_ids: set[str] = set()
    for link in links:
        incident_ids.add(link["source"] if isinstance(link["source"], str) else link["source"])
        incident_ids.add(link["target"] if isinstance(link["target"], str) else link["target"])
    orphans = [n for n in nodes if n["id"] not in incident_ids]
    if orphans:
        orphan_ids = [n["id"] for n in orphans]
        print(f"Warning: {len(orphans)} orphan nodes found: {orphan_ids}")

    return {
        "nodes": nodes,
        "links": links,
    }


def extract_narrative(rdf_path: str | Path, base_iri: str) -> dict:
    """Extract narrative sections (FAQ, glossary, HowTo, People, Orgs) from RDF."""
    g = Graph()
    g.parse(str(rdf_path))

    result = {
        "faq": [],
        "glossary": [],
        "howto": [],
        "people": [],
        "organizations": [],
    }

    # FAQ
    for faq in g.subjects(RDF.type, SCHEMA.FAQPage):
        for q_item in g.objects(faq, SCHEMA.hasPart):
            q_text = extract_label(q_item, g) if isinstance(q_item, URIRef) else str(q_item)
            a_iri = None
            for a_item in g.objects(q_item, SCHEMA.acceptedAnswer):
                a_iri = a_item
                break
            a_text = ""
            for txt in g.objects(a_item, SCHEMA.text):
                a_text = str(txt)
                break
            for txt in g.objects(a_item, RDFS.comment):
                if not a_text:
                    a_text = str(txt)
                break
            if q_text and a_text:
                result["faq"].append({
                    "question": q_text,
                    "answer": a_text,
                    "iri": str(q_item) if isinstance(q_item, URIRef) else "",
                })

    # Fallback: look for Question nodes directly
    if not result["faq"]:
        for q in g.subjects(RDF.type, SCHEMA.Question):
            q_text = extract_label(q, g) or ""
            for a in g.objects(q, SCHEMA.acceptedAnswer):
                a_text = ""
                for txt in g.objects(a, SCHEMA.text):
                    a_text = str(txt)
                    break
                if q_text and a_text:
                    result["faq"].append({
                        "question": q_text,
                        "answer": a_text,
                        "iri": str(q) if isinstance(q, URIRef) else "",
                    })

    # Glossary
    for term_set in g.subjects(RDF.type, SCHEMA.DefinedTermSet):
        for term in g.objects(term_set, SCHEMA.hasPart):
            term_text = extract_label(term, g) if isinstance(term, URIRef) else str(term)
            term_desc = extract_description(term, g) if isinstance(term, URIRef) else ""
            if term_text and term_desc:
                result["glossary"].append({
                    "term": term_text,
                    "definition": term_desc,
                    "iri": str(term) if isinstance(term, URIRef) else "",
                })

    # Fallback glossary: DefinedTerm nodes
    if not result["glossary"]:
        for term in g.subjects(RDF.type, SCHEMA.DefinedTerm):
            term_text = extract_label(term, g) or ""
            desc = extract_description(term, g) or ""
            if term_text:
                result["glossary"].append({
                    "term": term_text,
                    "definition": desc,
                    "iri": str(term) if isinstance(term, URIRef) else "",
                })

    # HowTo
    for howto in g.subjects(RDF.type, SCHEMA.HowTo):
        for step in g.objects(howto, SCHEMA.step):
            step_text = extract_label(step, g) if isinstance(step, URIRef) else str(step)
            step_desc = extract_description(step, g) if isinstance(step, URIRef) else ""
            if step_text:
                result["howto"].append({
                    "step": step_text,
                    "description": step_desc,
                    "iri": str(step) if isinstance(step, URIRef) else "",
                })

    # People
    for person in g.subjects(RDF.type, SCHEMA.Person):
        name = extract_label(person, g)
        if name:
            desc = extract_description(person, g)
            result["people"].append({
                "name": name,
                "description": desc,
                "iri": str(person) if isinstance(person, URIRef) else "",
            })

    # Organizations
    for org in g.subjects(RDF.type, SCHEMA.Organization):
        name = extract_label(org, g)
        if name:
            desc = extract_description(org, g)
            result["organizations"].append({
                "name": name,
                "description": desc,
                "iri": str(org) if isinstance(org, URIRef) else "",
            })

    return result


def get_base_iri(rdf_path: str | Path) -> str:
    """Extract the base IRI from an RDF file if available."""
    g = Graph()
    g.parse(str(rdf_path))
    for s in set(g.subjects()):
        if isinstance(s, URIRef):
            uri = str(s)
            if "#" in uri:
                return uri.split("#")[0] + "#"
            return uri.rsplit("/", 1)[0] + "/"
    return "https://linkedin.com/pulse/"


def get_entity_count(rdf_path: str | Path) -> int:
    """Return the number of triples in the RDF file."""
    g = Graph()
    g.parse(str(rdf_path))
    return len(g)


def validate_orphans(kgdata: dict) -> list[str]:
    """Return list of orphan node IDs (nodes with no incident links)."""
    incident: set[str] = set()
    for link in kgdata["links"]:
        src = link["source"] if isinstance(link["source"], str) else link["source"]["id"]
        tgt = link["target"] if isinstance(link["target"], str) else link["target"]["id"]
        incident.add(src)
        incident.add(tgt)
    orphans = [n["id"] for n in kgdata["nodes"] if n["id"] not in incident]
    return orphans

File: rdf-infographic-skill/scripts/rdf_parser.ts
/**
 * RDF Parser — TypeScript edition (Node.js ≥ 18, requires n3).
 * Consolidates rdf_parser.py (importable) and rdf-parser.py (CLI) into one file.
 *
 * Install: npm install (from rdf-infographic-skill/scripts/)
 * Run CLI: npx tsx rdf_parser.ts input.ttl [--format turtle|nt|n3]
 */

import { readFileSync } from "node:fs";
import { DataFactory, Parser, Store, type Quad_Object, type Quad_Subject, type Term } from "n3";

const { namedNode } = DataFactory;

// ── Well-known IRIs ───────────────────────────────────────────────────────────

const RDF_TYPE      = namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type");
const RDFS_LABEL    = namedNode("http://www.w3.org/2000/01/rdf-schema#label");
const RDFS_COMMENT  = namedNode("http://www.w3.org/2000/01/rdf-schema#comment");
const RDFS_CLASS    = namedNode("http://www.w3.org/2000/01/rdf-schema#Class");
const OWL_CLASS     = namedNode("http://www.w3.org/2002/07/owl#Class");
const OWL_PROPERTY  = namedNode("http://www.w3.org/2002/07/owl#ObjectProperty");
const RDF_PROPERTY  = namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#Property");

const SCHEMA_NAME        = namedNode("http://schema.org/name");
const SCHEMA_DESCRIPTION = namedNode("http://schema.org/description");
const SCHEMA_PERSON      = namedNode("http://schema.org/Person");
const SCHEMA_ORG         = namedNode("http://schema.org/Organization");
const SCHEMA_ARTICLE     = namedNode("http://schema.org/Article");
const SCHEMA_FAQPAGE     = namedNode("http://schema.org/FAQPage");
const SCHEMA_QUESTION    = namedNode("http://schema.org/Question");
const SCHEMA_HOWTO       = namedNode("http://schema.org/HowTo");
const SCHEMA_HOWTO_STEP  = namedNode("http://schema.org/HowToStep");
const SCHEMA_DEFINED_TERM_SET = namedNode("http://schema.org/DefinedTermSet");
const SCHEMA_DEFINED_TERM    = namedNode("http://schema.org/DefinedTerm");
const SCHEMA_HAS_PART    = namedNode("http://schema.org/hasPart");
const SCHEMA_STEP        = namedNode("http://schema.org/step");
const SCHEMA_ACCEPTED_ANSWER = namedNode("http://schema.org/acceptedAnswer");
const SCHEMA_TEXT        = namedNode("http://schema.org/text");
const SCHEMA_CREATIVE_WORK = namedNode("http://schema.org/CreativeWork");

const KNOWN_CLASS_IRIS = new Set([
  RDF_PROPERTY.value, RDFS_CLASS.value, OWL_CLASS.value,
  SCHEMA_PERSON.value, SCHEMA_ORG.value, SCHEMA_ARTICLE.value,
  SCHEMA_FAQPAGE.value, SCHEMA_QUESTION.value,
  SCHEMA_DEFINED_TERM_SET.value, SCHEMA_DEFINED_TERM.value,
  SCHEMA_HOWTO.value, SCHEMA_HOWTO_STEP.value,
  "http://schema.org/SoftwareApplication", "http://schema.org/SoftwareSourceCode",
  "http://schema.org/Thing", SCHEMA_CREATIVE_WORK.value,
]);

// ── Types ─────────────────────────────────────────────────────────────────────

export interface KgNode {
  id: string;
  group: string;
  label: string;
  desc: string;
  iri: string;
}

export interface KgLink {
  source: string;
  target: string;
  predicate: string;
  label: string;
}

export interface KgData {
  nodes: KgNode[];
  links: KgLink[];
}

export interface NarrativeEntry {
  name?: string;
  question?: string;
  term?: string;
  step?: string;
  answer?: string;
  definition?: string;
  description?: string;
  iri: string;
}

export interface NarrativeData {
  faq: Array<{ question: string; answer: string; iri: string }>;
  glossary: Array<{ term: string; definition: string; iri: string }>;
  howto: Array<{ step: string; description: string; iri: string }>;
  people: Array<{ name: string; description: string; iri: string }>;
  organizations: Array<{ name: string; description: string; iri: string }>;
}

// ── Parse helpers ─────────────────────────────────────────────────────────────

/** Parse an RDF file and return a populated n3 Store + prefix map. */
export function parseFile(rdfPath: string, format?: string): { store: Store; prefixes: Map<string, string> } {
  const content = readFileSync(rdfPath, "utf-8");
  return parseText(content, format ?? autoFormat(rdfPath));
}

export function parseText(content: string, format = "Turtle"): { store: Store; prefixes: Map<string, string> } {
  const store = new Store();
  const prefixes = new Map<string, string>();
  // n3 Parser is synchronous when given a string
  const parser = new Parser({ format: format as Parameters<typeof Parser>[0]["format"] });
  const quads = parser.parse(content, null, (prefix, ns) => {
    prefixes.set(prefix, ns.value);
  });
  store.addQuads(quads);
  return { store, prefixes };
}

function autoFormat(path: string): string {
  const ext = path.toLowerCase().split(".").pop() ?? "";
  return ({ ttl: "Turtle", n3: "N3", nt: "N-Triples", rdf: "application/rdf+xml", xml: "application/rdf+xml" } as Record<string, string>)[ext] ?? "Turtle";
}

// ── Label / description helpers ───────────────────────────────────────────────

function firstLiteral(store: Store, subject: Term, ...predicates: Term[]): string {
  for (const pred of predicates) {
    for (const obj of store.getObjects(subject, pred, null)) {
      if (obj.termType === "Literal") return obj.value;
    }
  }
  return "";
}

function shortenUri(uri: string, prefixes: Map<string, string>): string {
  for (const [prefix, ns] of prefixes) {
    if (uri.startsWith(ns)) return `${prefix}:${uri.slice(ns.length)}`;
  }
  if (uri.includes("#")) return uri.split("#").pop() ?? uri;
  return uri.split("/").pop() ?? uri;
}

function extractLabel(uri: string, store: Store, prefixes: Map<string, string>): string {
  const node = namedNode(uri);
  return firstLiteral(store, node, RDFS_LABEL, SCHEMA_NAME) || shortenUri(uri, prefixes);
}

function extractDescription(uri: string, store: Store): string {
  const node = namedNode(uri);
  const desc = firstLiteral(store, node, RDFS_COMMENT, SCHEMA_DESCRIPTION);
  return desc.slice(0, 200);
}

function classifyNode(uri: string, store: Store): string {
  const node = namedNode(uri);
  const types = store.getObjects(node, RDF_TYPE, null).map(t => t.value);
  if (types.includes(RDFS_CLASS.value) || types.includes(OWL_CLASS.value)) return "Class";
  if (types.includes(RDF_PROPERTY.value) || types.includes(OWL_PROPERTY.value)) return "Property";
  // Check if used as predicate
  if (store.countQuads(null, node, null, null) > 0) return "Property";
  // Check if used as a class (something has rdf:type pointing to it)
  if (store.countQuads(null, RDF_TYPE, node, null) > 0) return "Class";
  // Check known class IRIs
  if (KNOWN_CLASS_IRIS.has(uri)) return "Instance";
  return "Instance";
}

// ── buildKgdata ───────────────────────────────────────────────────────────────

export function buildKgdata(rdfPath: string): KgData {
  const { store, prefixes } = parseFile(rdfPath);
  const nodesMap = new Map<string, KgNode>();
  const links: KgLink[] = [];

  for (const quad of store) {
    const { subject: s, predicate: p, object: o } = quad;
    if (s.termType === "BlankNode") continue;
    if (o.termType === "BlankNode") continue;

    const predShort = p.termType === "NamedNode" ? shortenUri(p.value, prefixes) : p.value;
    const subjId = s.termType === "NamedNode" ? s.value : `_:${(s as Term).value}`;
    const objId  = o.termType === "NamedNode" ? o.value : `_:${(o as Term).value}`;

    if (s.termType === "NamedNode" && !nodesMap.has(subjId)) {
      nodesMap.set(subjId, {
        id: subjId, group: classifyNode(subjId, store),
        label: extractLabel(subjId, store, prefixes),
        desc: extractDescription(subjId, store), iri: subjId,
      });
    }
    if (o.termType === "NamedNode" && !nodesMap.has(objId)) {
      nodesMap.set(objId, {
        id: objId, group: classifyNode(objId, store),
        label: extractLabel(objId, store, prefixes),
        desc: extractDescription(objId, store), iri: objId,
      });
    }
    if (p.termType === "NamedNode" && s.termType !== "Literal" && o.termType !== "Literal") {
      links.push({ source: subjId, target: objId, predicate: predShort, label: predShort });
    }
  }

  const nodes = [...nodesMap.values()];
  const incidentIds = new Set(links.flatMap(l => [l.source, l.target]));
  const orphans = nodes.filter(n => !incidentIds.has(n.id));
  if (orphans.length) {
    process.stderr.write(`Warning: ${orphans.length} orphan nodes found\n`);
  }

  return { nodes, links };
}

// ── extractNarrative ──────────────────────────────────────────────────────────

export function extractNarrative(rdfPath: string, _baseIri: string): NarrativeData {
  const { store, prefixes } = parseFile(rdfPath);
  const result: NarrativeData = { faq: [], glossary: [], howto: [], people: [], organizations: [] };

  // FAQ via FAQPage
  for (const faq of store.getSubjects(RDF_TYPE, SCHEMA_FAQPAGE, null)) {
    for (const qItem of store.getObjects(faq, SCHEMA_HAS_PART, null)) {
      if (qItem.termType !== "NamedNode") continue;
      const qText = extractLabel(qItem.value, store, prefixes);
      const aItems = store.getObjects(qItem, SCHEMA_ACCEPTED_ANSWER, null);
      for (const aItem of aItems) {
        const aText = firstLiteral(store, aItem, SCHEMA_TEXT, RDFS_COMMENT);
        if (qText && aText) {
          result.faq.push({ question: qText, answer: aText, iri: qItem.value });
        }
        break;
      }
    }
  }

  // FAQ fallback: direct Question nodes
  if (!result.faq.length) {
    for (const q of store.getSubjects(RDF_TYPE, SCHEMA_QUESTION, null)) {
      if (q.termType !== "NamedNode") continue;
      const qText = extractLabel(q.value, store, prefixes);
      for (const a of store.getObjects(q, SCHEMA_ACCEPTED_ANSWER, null)) {
        const aText = firstLiteral(store, a, SCHEMA_TEXT, RDFS_COMMENT);
        if (qText && aText) result.faq.push({ question: qText, answer: aText, iri: q.value });
        break;
      }
    }
  }

  // Glossary via DefinedTermSet
  for (const ts of store.getSubjects(RDF_TYPE, SCHEMA_DEFINED_TERM_SET, null)) {
    for (const term of store.getObjects(ts, SCHEMA_HAS_PART, null)) {
      if (term.termType !== "NamedNode") continue;
      const termText = extractLabel(term.value, store, prefixes);
      const termDesc = extractDescription(term.value, store);
      if (termText && termDesc) result.glossary.push({ term: termText, definition: termDesc, iri: term.value });
    }
  }

  // Glossary fallback: direct DefinedTerm nodes
  if (!result.glossary.length) {
    for (const term of store.getSubjects(RDF_TYPE, SCHEMA_DEFINED_TERM, null)) {
      if (term.termType !== "NamedNode") continue;
      const termText = extractLabel(term.value, store, prefixes);
      const termDesc = extractDescription(term.value, store);
      if (termText) result.glossary.push({ term: termText, definition: termDesc, iri: term.value });
    }
  }

  // HowTo
  for (const howto of store.getSubjects(RDF_TYPE, SCHEMA_HOWTO, null)) {
    for (const step of store.getObjects(howto, SCHEMA_STEP, null)) {
      if (step.termType !== "NamedNode") continue;
      const stepText = extractLabel(step.value, store, prefixes);
      const stepDesc = extractDescription(step.value, store);
      if (stepText) result.howto.push({ step: stepText, description: stepDesc, iri: step.value });
    }
  }

  // People
  for (const person of store.getSubjects(RDF_TYPE, SCHEMA_PERSON, null)) {
    if (person.termType !== "NamedNode") continue;
    const name = extractLabel(person.value, store, prefixes);
    if (name) result.people.push({ name, description: extractDescription(person.value, store), iri: person.value });
  }

  // Organizations
  for (const org of store.getSubjects(RDF_TYPE, SCHEMA_ORG, null)) {
    if (org.termType !== "NamedNode") continue;
    const name = extractLabel(org.value, store, prefixes);
    if (name) result.organizations.push({ name, description: extractDescription(org.value, store), iri: org.value });
  }

  return result;
}

// ── Utility exports ───────────────────────────────────────────────────────────

export function getBaseIri(rdfPath: string): string {
  const { store } = parseFile(rdfPath);
  for (const subject of store.getSubjects(null, null, null)) {
    if (subject.termType === "NamedNode") {
      const uri = subject.value;
      if (uri.includes("#")) return uri.split("#")[0] + "#";
      return uri.replace(/\/[^/]*$/, "/");
    }
  }
  return "https://linkedin.com/pulse/";
}

export function getEntityCount(rdfPath: string): number {
  const { store } = parseFile(rdfPath);
  return store.size;
}

export function validateOrphans(kgdata: KgData): string[] {
  const incident = new Set(kgdata.links.flatMap(l => [
    typeof l.source === "string" ? l.source : (l.source as { id: string }).id,
    typeof l.target === "string" ? l.target : (l.target as { id: string }).id,
  ]));
  return kgdata.nodes.filter(n => !incident.has(n.id)).map(n => n.id);
}

// ── CLI (mirrors rdf-parser.py) ───────────────────────────────────────────────

if (
  process.argv[1] &&
  (process.argv[1].endsWith("rdf_parser.ts") || process.argv[1].endsWith("rdf_parser.js"))
) {
  const argv = process.argv.slice(2);
  if (!argv.length) {
    process.stderr.write("Usage: npx tsx rdf_parser.ts <input.ttl> [--format turtle|nt|n3]\n");
    process.exit(1);
  }
  const inputFile = argv[0];
  let format = "Turtle";
  const fmtIdx = argv.indexOf("--format");
  if (fmtIdx >= 0 && argv[fmtIdx + 1]) format = argv[fmtIdx + 1];

  try {
    const kgdata = buildKgdata(inputFile);
    const { store, prefixes } = parseFile(inputFile, format);
    const entityTypes: Record<string, string[]> = {};
    for (const quad of store.match(null, RDF_TYPE, null, null)) {
      const t = quad.object.value;
      (entityTypes[t] ??= []).push(quad.subject.value);
    }
    const config = {
      ENTITY_TYPES: Object.keys(entityTypes).slice(0, 10),
      ENTITIES_COUNT: kgdata.nodes.length,
      RELATIONSHIPS_COUNT: kgdata.links.length,
    };
    process.stdout.write(JSON.stringify(config, null, 2) + "\n");
  } catch (err) {
    process.stderr.write(`Error: ${(err as Error).message}\n`);
    process.exit(1);
  }
}
